{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "057bb63c",
   "metadata": {},
   "source": [
    "# Cell type assignment\n",
    "\n",
    "In this notebook, we analyse the cell type predictions of the automated annotation model and validate its cell type assignments using marker gene expression and other sources of information (e.g. knowledge of developmental stages and anatomical dissections).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "c2089d2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scrabbitpy\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "import matplotlib.pyplot as plt\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72d96d6d",
   "metadata": {},
   "source": [
    "### Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "9cb1f1c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load data\n",
    "r_data = sc.read_h5ad(\"../data-in/cell_type_annotation/rabbit_corrected_clustered.h5ad\")\n",
    "orthologs = pd.read_csv(\"../data-in/orthologs/rabbit_mouse_orthologs.tsv\", sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "664103fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "m_markers = pd.read_csv(\"../data-in/cell_type_annotation/celltype_markers.tsv\", sep=\"\\t\")\n",
    "m_markers = {i: m_markers[i].dropna().to_list() for i in m_markers.columns}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d5374e0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a3f589f0",
   "metadata": {},
   "source": [
    "### Cluster data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20daf5ca",
   "metadata": {},
   "source": [
    "Cell type labels are assigned to clusters identified in the high-dimensional gene expression space. Here we perfrom Leiden clustering at various resolutions to identify both coarse and fine-grained populations of cells.\n",
    "\n",
    "For the sake of brevity, here we have used precomputed clusterings. See `scripts/cluster.py` to run the full clustering.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9efc2a39",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load clusters from file\n",
    "clusters = pd.read_csv(\"../data-in/cell_type_annotation/clusters.tsv\",sep=\"\\t\")\n",
    "r_data.obs = r_data.obs.join(clusters)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4eddce91",
   "metadata": {},
   "source": [
    "### Divide dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60a3f171",
   "metadata": {},
   "source": [
    "To simplify the annotation process, the dataset is initially divded into broad regions to be annotated independently. \n",
    "\n",
    "For each broad region, we plot the automated annotation predictions, along with clusterings of different resolutions. In addition to clustering the entire dataset, Leiden clustering is also performed within each region. These results are compared to UMAP plots of marker gene expression of known cell types as well as differentially expressed genes computationally identified in the annotated mouse dataset.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2eafe765",
   "metadata": {},
   "outputs": [],
   "source": [
    "blood = rabbit[rabbit.obs[\"leiden_res1\"].isin(['4','19','20','8']),]\n",
    "mesoderm = rabbit[rabbit.obs[\"leiden_res1\"].isin(['17','2','12','22']),]\n",
    "neural = rabbit[rabbit.obs[\"leiden_res1\"].isin(['3','7']),]\n",
    "misc = rabbit[rabbit.obs[\"leiden_res1\"].isin(['6','23','21','18']),]\n",
    "exe_ectoderm = rabbit[rabbit.obs[\"leiden_res1\"].isin(['9','0','16']),]\n",
    "exe_mesoderm = rabbit[rabbit.obs[\"leiden_res1\"].isin(['5','14','10','1','22']),]\n",
    "exe_endoderm = rabbit[rabbit.obs[\"leiden_res1\"].isin(['13','15','11']),]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "ffdb0864",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create directory for each region\n",
    "# Identify mouse cell types likely to be in each \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "c0410a85",
   "metadata": {},
   "outputs": [],
   "source": [
    "def createAnnotationDirs(base_path, clusters):\n",
    "    \"\"\"\n",
    "    \n",
    "    e.g.\n",
    "    createAnnotationDirs(base_path = \"../figs/celltype_annotation/annotation_pipeline/r_mesoderm/\", \n",
    "                     clusters = [\"leiden_res8\",\"mesoderm_leiden_res2\",\"mesoderm_leiden_res5\"])\n",
    "                     \n",
    "    \"\"\"\n",
    "    \n",
    "    # Marker expression plots\n",
    "    Path(base_path + \"marker_expression\").mkdir(parents=True, exist_ok=True)\n",
    "    \n",
    "    # Cluster fraction plots\n",
    "    Path(base_path + \"cluster_fractions\").mkdir(parents=True, exist_ok=True)\n",
    "    \n",
    "    # Prediction fraction plots\n",
    "    Path(base_path + \"cluster_fractions\").mkdir(parents=True, exist_ok=True)\n",
    "    \n",
    "    for cluster in clusters:\n",
    "        # Create directory for each clustering\n",
    "        Path(base_path + \"cluster_fractions/\" + cluster).mkdir(parents=True, exist_ok=True)\n",
    "        Path(base_path + \"prediction_fractions/\" + cluster).mkdir(parents=True, exist_ok=True)\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d6f968f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "631b1339",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28155bf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plotMarkerExpression(adata, markers,):\n",
    "    \"\"\"\n",
    "    adata: Scanpy object\n",
    "    markers: Dictionary where keys are the cell types/obs groups of interest and items are marker gene names which match values in adata.var.index.   \n",
    "    \"\"\"\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "161d19d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def computeObsFraction(adata, obsA=\"leiden_res8\",obsB=\"singler\", obsB_ignore_thresh = None ):\n",
    "    \"\"\"\n",
    "    Computes the fraction of observations B within groups of observation A.\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    df = pd.DataFrame({obsA:adata.obs[obsA], obsB:adata.obs[obsB]},index=adata.obs.index)\n",
    "    df[obsA + \"_ncells\"] = df.groupby([obsA]).transform('count')\n",
    "\n",
    "    df[obsB + \"_count\"] = 0\n",
    "\n",
    "    counts = df.groupby([obsA,obsB,obsA + \"_ncells\"]).count()\n",
    "    counts[obsB + \"_frac\"] = counts[obsB + \"_count\"].groupby(obsA).transform(lambda x: x/x.sum())\n",
    "    counts.sort_values([obsA,obsB,obsB + \"_frac\"],ascending=False)\n",
    "    \n",
    "    counts= counts.reset_index()\n",
    "    counts = counts.dropna()\n",
    "    \n",
    "     # Group observations B with very few cells as 'Other'\n",
    "    if(obsB_ignore_thresh is not None) :\n",
    "        counts[obsB] = counts[obsB].astype(str)\n",
    "        counts.loc[counts[obsB + \"_frac\"] <= obsB_ignore_thresh,obsB] = \"Other\"      \n",
    "        \n",
    "    \n",
    "    \n",
    "    return(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "cf967cad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>leiden_res8</th>\n",
       "      <th>singler</th>\n",
       "      <th>leiden_res8_ncells</th>\n",
       "      <th>singler_count</th>\n",
       "      <th>singler_frac</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1066</th>\n",
       "      <td>3</td>\n",
       "      <td>Non-neural ectoderm 4</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1931</th>\n",
       "      <td>6</td>\n",
       "      <td>Caudal mesoderm</td>\n",
       "      <td>37</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.513514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2243</th>\n",
       "      <td>6</td>\n",
       "      <td>Other</td>\n",
       "      <td>37</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.027027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2425</th>\n",
       "      <td>6</td>\n",
       "      <td>Lateral plate mesoderm</td>\n",
       "      <td>37</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.216216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3023</th>\n",
       "      <td>6</td>\n",
       "      <td>Presomitic mesoderm</td>\n",
       "      <td>37</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.243243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47306</th>\n",
       "      <td>79</td>\n",
       "      <td>Sclerotome</td>\n",
       "      <td>294</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.085034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47332</th>\n",
       "      <td>79</td>\n",
       "      <td>Other</td>\n",
       "      <td>294</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.017007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47358</th>\n",
       "      <td>79</td>\n",
       "      <td>Other</td>\n",
       "      <td>294</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.030612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47410</th>\n",
       "      <td>79</td>\n",
       "      <td>Other</td>\n",
       "      <td>294</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.003401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47436</th>\n",
       "      <td>79</td>\n",
       "      <td>Other</td>\n",
       "      <td>294</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.027211</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>448 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      leiden_res8                 singler  leiden_res8_ncells  singler_count  \\\n",
       "1066            3   Non-neural ectoderm 4                   1            1.0   \n",
       "1931            6         Caudal mesoderm                  37           19.0   \n",
       "2243            6                   Other                  37            1.0   \n",
       "2425            6  Lateral plate mesoderm                  37            8.0   \n",
       "3023            6     Presomitic mesoderm                  37            9.0   \n",
       "...           ...                     ...                 ...            ...   \n",
       "47306          79              Sclerotome                 294           25.0   \n",
       "47332          79                   Other                 294            5.0   \n",
       "47358          79                   Other                 294            9.0   \n",
       "47410          79                   Other                 294            1.0   \n",
       "47436          79                   Other                 294            8.0   \n",
       "\n",
       "       singler_frac  \n",
       "1066       1.000000  \n",
       "1931       0.513514  \n",
       "2243       0.027027  \n",
       "2425       0.216216  \n",
       "3023       0.243243  \n",
       "...             ...  \n",
       "47306      0.085034  \n",
       "47332      0.017007  \n",
       "47358      0.030612  \n",
       "47410      0.003401  \n",
       "47436      0.027211  \n",
       "\n",
       "[448 rows x 5 columns]"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = computeObsFraction(r_mesoderm, obsA = \"leiden_res8\", obsB = \"singler\", obsB_ignore_thresh = 0.05)\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "0b617c39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>leiden_res8</th>\n",
       "      <th>singler</th>\n",
       "      <th>leiden_res8_ncells</th>\n",
       "      <th>singler_count</th>\n",
       "      <th>singler_frac</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6587</th>\n",
       "      <td>11</td>\n",
       "      <td>Other</td>\n",
       "      <td>64</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.046875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6639</th>\n",
       "      <td>11</td>\n",
       "      <td>Other</td>\n",
       "      <td>64</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.015625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6743</th>\n",
       "      <td>11</td>\n",
       "      <td>Other</td>\n",
       "      <td>64</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.015625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6769</th>\n",
       "      <td>11</td>\n",
       "      <td>Cardiopharyngeal progenitors FHF</td>\n",
       "      <td>64</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.140625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6795</th>\n",
       "      <td>11</td>\n",
       "      <td>Cardiopharyngeal progenitors SHF</td>\n",
       "      <td>64</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.062500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47306</th>\n",
       "      <td>79</td>\n",
       "      <td>Sclerotome</td>\n",
       "      <td>294</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.085034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47332</th>\n",
       "      <td>79</td>\n",
       "      <td>Other</td>\n",
       "      <td>294</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.017007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47358</th>\n",
       "      <td>79</td>\n",
       "      <td>Other</td>\n",
       "      <td>294</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.030612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47410</th>\n",
       "      <td>79</td>\n",
       "      <td>Other</td>\n",
       "      <td>294</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.003401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47436</th>\n",
       "      <td>79</td>\n",
       "      <td>Other</td>\n",
       "      <td>294</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.027211</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>395 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      leiden_res8                           singler  leiden_res8_ncells  \\\n",
       "6587           11                             Other                  64   \n",
       "6639           11                             Other                  64   \n",
       "6743           11                             Other                  64   \n",
       "6769           11  Cardiopharyngeal progenitors FHF                  64   \n",
       "6795           11  Cardiopharyngeal progenitors SHF                  64   \n",
       "...           ...                               ...                 ...   \n",
       "47306          79                        Sclerotome                 294   \n",
       "47332          79                             Other                 294   \n",
       "47358          79                             Other                 294   \n",
       "47410          79                             Other                 294   \n",
       "47436          79                             Other                 294   \n",
       "\n",
       "       singler_count  singler_frac  \n",
       "6587             3.0      0.046875  \n",
       "6639             1.0      0.015625  \n",
       "6743             1.0      0.015625  \n",
       "6769             9.0      0.140625  \n",
       "6795             4.0      0.062500  \n",
       "...              ...           ...  \n",
       "47306           25.0      0.085034  \n",
       "47332            5.0      0.017007  \n",
       "47358            9.0      0.030612  \n",
       "47410            1.0      0.003401  \n",
       "47436            8.0      0.027211  \n",
       "\n",
       "[395 rows x 5 columns]"
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.loc[data[obsA + \"_ncells\"] >= 50]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "34cd71b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([], dtype=int64),)"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncells = df.groupby([obsA,obsA + \"_ncells\"]).count()\n",
    "np.where(ncells[obsB] <= 0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "1f325727",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>leiden_res8</th>\n",
       "      <th>singler</th>\n",
       "      <th>leiden_res8_ncells</th>\n",
       "      <th>singler_count</th>\n",
       "      <th>singler_frac</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1066</th>\n",
       "      <td>3</td>\n",
       "      <td>Non-neural ectoderm 4</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1905</th>\n",
       "      <td>6</td>\n",
       "      <td>Caudal mesoderm</td>\n",
       "      <td>37</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.513514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2217</th>\n",
       "      <td>6</td>\n",
       "      <td>Other</td>\n",
       "      <td>37</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.027027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2399</th>\n",
       "      <td>6</td>\n",
       "      <td>Lateral plate mesoderm</td>\n",
       "      <td>37</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.216216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2997</th>\n",
       "      <td>6</td>\n",
       "      <td>Presomitic mesoderm</td>\n",
       "      <td>37</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.243243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46578</th>\n",
       "      <td>79</td>\n",
       "      <td>Sclerotome</td>\n",
       "      <td>294</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.085034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46604</th>\n",
       "      <td>79</td>\n",
       "      <td>Other</td>\n",
       "      <td>294</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.017007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46630</th>\n",
       "      <td>79</td>\n",
       "      <td>Other</td>\n",
       "      <td>294</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.030612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46682</th>\n",
       "      <td>79</td>\n",
       "      <td>Other</td>\n",
       "      <td>294</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.003401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46708</th>\n",
       "      <td>79</td>\n",
       "      <td>Other</td>\n",
       "      <td>294</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.027211</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>448 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      leiden_res8                 singler  leiden_res8_ncells  singler_count  \\\n",
       "1066            3   Non-neural ectoderm 4                   1            1.0   \n",
       "1905            6         Caudal mesoderm                  37           19.0   \n",
       "2217            6                   Other                  37            1.0   \n",
       "2399            6  Lateral plate mesoderm                  37            8.0   \n",
       "2997            6     Presomitic mesoderm                  37            9.0   \n",
       "...           ...                     ...                 ...            ...   \n",
       "46578          79              Sclerotome                 294           25.0   \n",
       "46604          79                   Other                 294            5.0   \n",
       "46630          79                   Other                 294            9.0   \n",
       "46682          79                   Other                 294            1.0   \n",
       "46708          79                   Other                 294            8.0   \n",
       "\n",
       "       singler_frac  \n",
       "1066       1.000000  \n",
       "1905       0.513514  \n",
       "2217       0.027027  \n",
       "2399       0.216216  \n",
       "2997       0.243243  \n",
       "...             ...  \n",
       "46578      0.085034  \n",
       "46604      0.017007  \n",
       "46630      0.030612  \n",
       "46682      0.003401  \n",
       "46708      0.027211  \n",
       "\n",
       "[448 rows x 5 columns]"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#test[obsB].cat.add_categories(\"Other\", inplace=True)\n",
    "#test.loc[test[obsB + \"_frac\"] <= 0.05,obsB] = \"Other\"\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8e57584",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plotObsFraction()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "f2ab02f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plotMultiObsFraction(data, obsA, obsB, obsA_ignore_thresh = None, obsB_ignore_thresh = None, export_dir=None, obsB_colours=None):\n",
    "    \n",
    "    \n",
    "    # Get obs fractions if anndata object passed in\n",
    "    if(isinstance(r_mesoderm,sc.AnnData)):\n",
    "        data = computeObsFraction(data, obsA = obsA, obsB = obsB, obsB_ignore_thresh = obsB_ignore_thresh)\n",
    "    \n",
    "    # Don't plot fractions for observations with very few cells\n",
    "    if(obsA_ignore_thresh is not None):\n",
    "        data = data.loc[data[obsA + \"_ncells\"] >= obsA_ignore_thresh]\n",
    "    \n",
    "    \n",
    "    # if(export_dir is None):\n",
    "        # Plot in a grid\n",
    "    \n",
    "    \n",
    "    if(export_dir is not None):\n",
    "        for x in data[obsA].unique():\n",
    "            fig, ax = plt.subplots(figsize=(10,10))\n",
    "            \n",
    "            df_plot = data[data[obsA]==x]\n",
    "            \n",
    "            if(obsB_colours is not None):\n",
    "                obsB_colours = [obsB_colours[i] for i in df_plot[obsB]]\n",
    "                \n",
    "            ax.barh(df_plot[obsB], df_plot[obsB + \"_frac\"], color=obsB_colours)\n",
    "            \n",
    "            ax.set_title(obsA + \": \" +  x + \" (\" + str(int(df_plot[obsA + \"_ncells\"].iloc[0]))+ \" cells)\")\n",
    "            ax.set_xlabel(\"Fraction of cells\")\n",
    "            ax.set_ylabel(obsB)\n",
    "            \n",
    "            plt.tight_layout()\n",
    "            plt.savefig(export_dir + x + \".pdf\")\n",
    "            plt.close(fig)\n",
    "\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "255d9aff",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e77fa141",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ed56291",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "40bcb95e",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotMultiObsFraction(r_mesoderm, obsA = \"leiden_res8\", obsB = \"singler\", \n",
    "                     obsA_ignore_thresh = 50 , obsB_ignore_thresh = 0.02,\n",
    "                     \n",
    "                     #colours = scrabbit.getCelltypeColours(),\n",
    "                     export_dir = \"../figs/celltype_annotation/annotation_pipeline/r_mesoderm/cluster_fractions/leiden_res8/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "715f629b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def makeAnnotationPlots(adata, clusters, model_predictions, celltypes, markers, export_dir):\n",
    "    \n",
    "    \n",
    "    createAnnotationDirs(base_path = export_dir, clusters = clusters )\n",
    "    \n",
    "    \n",
    "    # Plot UMAPs of SingleR predictions, Clusterings\n",
    "    sc.settings.figdir = export_dir\n",
    "    sc.pl.umap(adata, color=[model_predictions,clusters],\n",
    "               save=\"_clusters.pdf\")\n",
    "    \n",
    "\n",
    "    # Plot UMAPs of literature marker genes\n",
    "    for celltype in celltypes:\n",
    "        sc.settings.figdir = export_dir + \"marker_expression/\"\n",
    "        sc.pl.umap(adata,color=markers[celltype],\n",
    "                   save=\"_\" + celltype + \"_markers.pdf\")\n",
    "    \n",
    "   \n",
    "    # Plot bar charts of cluster prediction fractions\n",
    "    for cluster in clusters:\n",
    "        sc.settings.figdir = export_dir + \"cluster_fractions/\" + cluster\n",
    "        \n",
    "        plotClusterFraction(adata, clusters=cluster, predictions=model_predictions, \n",
    "                            save= export_dir + \"cluster_fractions/\" + cluster)\n",
    "        \n",
    "    \n",
    "    # Plot bar charts of prediction cluster fractions\n",
    "    sc.settings.figdir = export_dir + \"cluster_fractions/\"\n",
    "        \n",
    "    plotClusterFraction(r_meosderm, clusters=\"leiden_res8\", predictions=\"singler\", save=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "486608e0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "1a4cfdaa",
   "metadata": {},
   "source": [
    "### Mesoderm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28b41c2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create directory for mesoderm region\n",
    "Path(\"../figs/celltype_annotation/annotation_pipeline/r_mesoderm/\").mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "runAnnotationPlots(r_mesoderm, \n",
    "                    clusters = [\"leiden_res8\",\"mesoderm_leiden_res2\",\"mesoderm_leiden_res5\"],\n",
    "                    model_predictions = \"singler\",\n",
    "                   marker_genes = m_markers,\n",
    "                   export_dir = \"../figs/celltype_annotation/annotation_pipeline/r_mesoderm/\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
