---
title: "Automated annotation"
output:
  html_document:
    df_print: paged
bibliography: references.bib
---

<br>

In this notebook, we leverage existing mouse gastrulation data and manually
annotated cell types to predict cell type annotations for the rabbit. 

Automated annotations approaches use the entire transcriptomic profile to 
predict cell types at a per-cell basis.

Here we used SingleR [@Aran2019], an automated annotation method which uses
the correlation across differentially expressed genes to determine cell type 
identities. Details of the method can be found 
[here](http://bioconductor.org/books/release/SingleRBook/). 
 
<br>

```{r,eval=FALSE}
library(scrabbitr)
library(SingleR)

set.seed(42)

out_path <- "../data-out/cell_type_annotation/"


```

```{r, eval=FALSE}

# Load data
r_data <- loadRabbitData()
m_data <- loadMouseData()
orthologs <- read.table("../data-in/orthologs.tsv",sep="\t")

```


```{r}
# Plot mouse cell types 
```



```{r,eval=FALSE}
model_train <- trainSingleR(ref = m_data[orthologs$mouse,],
                              labels = m_data$celltype,
                              de.method = "wilcox",
                              de.n = 50,
                              aggr.ref = TRUE)

saveRDS(model_train,paste0(out_path,"singler_mouse_trained.rds"))

```


```{r,eval=FALSE}

model_predict <- classifySingleR(test = r_data[orthologs$rabbit,],
                                 trained = model_train)

saveRDS(model_predict,paste0(out_path,"singler_rabbit_predicted.rds"))


```



```{r}
# Plot mouse and rabbit cell types side-by-side
```

<br>


# References

